### ISSUE TITLE
MPI_THREAD_MULTIPLE still broken in 1.8.5

### ISSUE THREAD

**Issue by gonnetp** [open-mpi/ompi#594]
I am currently working on a piece of software (SWIFT, see http://icc.dur.ac.uk/swift/) which relies heavily on MPI's asynchronous communication primitives and on `MPI_THREAD_MULTIPLE`.

Essentially what is happening is that the code will often call `MPI_Isend`, `MPI_Irecv`, and `MPI_Test` from more than one thread at the same time. The code works fine with Intel's MPI, IBM's Platform MPI, and with mpich as well, but it seems to hang quite reliably with OpenMPI 1.8.5.

The code hangs on a call to `MPI_Test`, i.e.

```
#0  __lll_unlock_wake ()
    at ../nptl/sysdeps/unix/sysv/linux/x86_64/lowlevellock.S:371
#1  0x00007ffff4c388a4 in _L_unlock_722 ()
   from /lib/x86_64-linux-gnu/libpthread.so.0
#2  0x00007ffff4c387f4 in __pthread_mutex_unlock_usercnt (decr=1, 
    mutex=0x603e0000b8c8) at pthread_mutex_unlock.c:57
#3  __GI___pthread_mutex_unlock (mutex=0x603e0000b8c8)
    at pthread_mutex_unlock.c:310
#4  0x00007fffeecb8c8d in mca_btl_vader_component_progress ()
   from /usr/lib/openmpi/mca_btl_vader.so
#5  0x00007ffff329257a in opal_progress () from /usr/lib/libopen-pal.so.6
#6  0x00007ffff3ffcbcb in ompi_request_default_test ()
   from /usr/lib/libmpi.so.1
#7  0x00007ffff402bee9 in PMPI_Test () from /usr/lib/libmpi.so.1
#8  0x00000000004be2b7 in task_lock (t=t@entry=0x7fffb9fce188) at task.c:111
#9  0x00000000004bd793 in queue_gettask (q=0x600c00024c20, 
    super=super@entry=0x7fffda400000, blocking=blocking@entry=0) at queue.c:203
#10 0x00000000004b005d in scheduler_gettask (s=s@entry=0x7fffffffdac0, qid=0, 
    super=super@entry=0x7fffda400000) at scheduler.c:1080
#11 0x0000000000479b11 in runner_main (data=0x600600036250) at runner.c:1253
#12 0x00007ffff4e63b98 in ?? () from /usr/lib/x86_64-linux-gnu/libasan.so.0
#13 0x00007ffff4c35182 in start_thread (arg=0x7fffd9b04700)
    at pthread_create.c:312
#14 0x00007ffff3adb47d in clone ()
    at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
```

The calls to `MPI_Isend` and `MPI_Irecv` that were executed concurrently all return normally. In this particular case, there was only a single thread calling `MPI_Test` while a single other thread called the send/recv functions.

The release notes to 1.8.5 say

> Fixed MPI_THREAD_MULTIPLE deadlock error in the vader BTL.  Thanks
> to Thomas Klimpel for reporting the issue.

The offending code seems to be in the same component. Could this be a similar, or even the same, issue in a slightly different form?

In order to reproduce the behaviour described above, download the latest version of SWIFT from https://gitlab.cosma.dur.ac.uk/swift/swiftsim and configure/build it with

```
./configure --with-gcc-arch=native  --with-metis --enable-mpi --enable-sanitizer --enable-debug=info
make
```

Where `--with-metis` and `--enable-sanitizer` are optional if you have neither Metis or `gcc-4.8` installed. To run the example do

```
cd examples/SodShock
python makeIC.py
cd ..
mpirun -np 2 xterm -e gdb -ex run --args ./test_fixdt_mpi -g "2 1 1" -r 1000 -t 1 -f SodShock/sodShock.hdf5 -m 0.01 -w 5000 -d 1e-4
```

and the code should hang after a few steps. On my laptop, each step takes about 5 seconds. If no output is generated for more than 10 seconds, then the code is stuck.

**Comment by hjelmn** [comment]
The backtrace looks like a normal iteration of the progress loop. This is a different issues. I will take a look now.

**Comment by hjelmn** [comment]
See the problem. There is a double-unlock in an error path. Fixing in master now.

**Comment by gonnetp** [comment]
Wow... That was quick!

Re-compiling from source, will check if this fixes everything.

Many, many thanks.

**Comment by hjelmn** [comment]
Looks like there is still an issue. Will commit a fix soon.

**Comment by hjelmn** [comment]
I think I get the remaining issue. The test appears to run fine on my mac.

**Comment by hjelmn** [comment]
Let me know if the issue is fixed. I will open a PR to bring the change into 1.8.6

**Comment by jsquyres** [comment]
Looks like this was fixed in 1.8.6.
